cosmwasm/packages/wasmi-runtime/src/wasm/runtime/contract.rs

```diff
@@ -469,8 +452,44 @@ impl WasmiApi for ContractInstance {
     }
 
     // stub, for now
-    fn query_chain_index(&mut self, _query_ptr_ptr: i32) -> Result<Option<RuntimeValue>, Trap> {
-        Err(WasmEngineError::NotImplemented.into())
+    fn query_chain_index(&mut self, query_ptr_ptr: i32) -> Result<Option<RuntimeValue>, Trap> {
+        let query_buffer = self.extract_vector(query_ptr_ptr as u32).map_err(|err| {
+            error!("query_chain() error while trying to read canonical address from wasm memory",);
+            err
+        })?;
+
+        trace!(
+            "query_chain() was called from WASM code with {:?}",
+            String::from_utf8_lossy(&query_buffer)
+        );
+
+        // TODO pass query_buffer to ocall
+        // Call query_chain (this bubbles up to x/compute via ocalls and FFI to Go code)
+        // Returns the value from x/compute
+        let (result, gas_used) = encrypt_and_query_chain(
+            &query_buffer,
+            &self.context,
+            self.user_nonce,
+            self.user_public_key,
+        )?;
+        self.use_gas_externally(gas_used)?;
+
+        let result = match result {
+            None => return Ok(Some(RuntimeValue::I32(0))), // Is this supposed to be 0 or Err?
```

Comment:
I think we should be returning errors in this case, but i'll check.

Reply:
Not an `Option` anymore, so this dilemma no longer exist.